### DISCUSSION TITLE
Some problem with Rcc config

### DISCUSSION THREAD

**Discussion by schrodinger1772** [stm32-rs/stm32f4xx-hal#815] [Q&A] [answered]
Hi,

When I want to config the clock for the system, I meet some problem. I used the stm32f401rct6 and the code was like let clock = rcc.cfgr
        .use_hse(8.MHz())
        .sysclk(84.MHz())
        .hclk(84.MHz())
        .pclk1(42.MHz())
        .pclk2(84.MHz())
        .freeze();
I copy that from the example, but that doesnot work when I use the usart1. But if I comment the sysclk, it can work but it's sysclk is 8MHZ.

If you know why that happen, I will appreciate that so much. Thanks for your time.

shvshy

**Comment by burrbull** [comment]
First of all I'd check the oscillator on your board. Try to comment out `.use_hse(8.MHz())` instead. Also there is usually 16 or 25 MHz external oscillators on F4 boards, not an 8 MHz.

**Comment by schrodinger1772** [comment]
Thanks a lot for your time. Here's the pic of the oscillator. 
![5b47dfc7b72209f30f02a1b40c21fe5](https://github.com/user-attachments/assets/75a55a66-a3d6-45b9-ae8f-40412b743914)
that's actually 8Mhz and I comment out the use_hse. there's nothing out from the usart1. Thanks again for you reply.

shevshy

**Comment by burrbull** [comment]
Is it freezes or it works but usart not working?
If freezes, try to find a place where it's happen. For example using `defmt::println`.

**Reply by schrodinger1772** [reply]
Hi，bother you again,

last nignt, there is a football game between chinese and Indonesia. I watch for that, sorry for my late. e, forgive my ignorance, I just try to use the defmt and i followed the instruction. But I got some errors below like error[E0599]: no function or associated item named `cycle_count` found for struct `DWT` in the current scope
   --> C:\Users\schroadinger\.cargo\registry\src\index.crates.io-6f17d22bba15001f\stm32f4xx-hal-0.20.0\src\dwt.rs:103:26.maybe I will also bother the defmt maintainer today.

so, I think you are really a master, so if there is a article for that. I am truely really sorry sorry for my questions, maybe just i am not that smart or have a great base.

thanks again!

shevshy

**Reply by burrbull** [reply]
Hmm.
Could you show the output of
```
cargo tree
```
Last error looks like dependency hell error.

Please insert code in "```" blocks. https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

**Reply by schrodinger1772** [reply]
Hi, that is kind of embarassing, I tick the run button on the vs of the defmt part, it works well, that's really useful. just forget my last reply, I am such a idiot. Thanks again. have a good day, I will try to find out if it freezed. 

shevshy

**Reply by schrodinger1772** [reply]
Hi, master.

Ohhh, that's really coventient using defmt, e, I found that clock has been freezed. and the problem maybe happened in let mut delay = dp.TIM1.delay_ms(&clock);that line. and I will give the error followd ERROR panicked at C:\Users\schroadinger\.cargo\registry\src\index.crates.io-6f17d22bba15001f\stm32f4xx-hal-0.22.0\src\timer.rs:780:55:
called `Result::unwrap()` on an `Err` value: TryFromIntError(())
└─ panic_probe::print_defmt::print @ C:\Users\schroadinger\.cargo\registry\src\index.crates.io-6f17d22bba15001f\panic-probe-0.3.2\src\lib.rs:104
Frame 0: syscall1 @ 0x08001fae inline
       C:\Users\schroadinger\.cargo\registry\src\index.crates.io-6f17d22bba15001f\cortex-m-semihosting-0.5.0\src/lib.rs:201:13
Frame 1: report_exception @ 0x0000000008001fac inline
       C:\Users\schroadinger\.cargo\registry\src\index.crates.io-6f17d22bba15001f\cortex-m-semihosting-0.5.0\src/macros.rs:28:9Frame 2: exit @ 0x0000000008001fac
       C:\Users\schroadinger\.cargo\registry\src\index.crates.io-6f17d22bba15001f\cortex-m-semihosting-0.5.0\src/debug.rs:73:25Frame 3: __cortex_m_rt_HardFault @ 0x08000834
       E:\Rust_32\my-app\src/lib.rs:36:9
Frame 4: HardFault_ @ 0x08003670
       C:\Users\schroadinger\.cargo\registry\src\index.crates.io-6f17d22bba15001f\cortex-m-rt-0.7.3\src/lib.rs:1046:5
Error: Semihosting indicates exit with reason: 0x20023

Stack backtrace:
   0: <unknown>
   1: <unknown>
   2: hid_write
   3: <unknown>
   4: <unknown>
   5: <unknown>
   6: <unknown>
   7: <unknown>
   8: <unknown>
   9: <unknown>
  10: hid_write
  11: BaseThreadInitThunk
  12: RtlUserThreadStart

when I changed the delay_ms as delay_us. everything goes well. So, what the problem of that. Thanks a lot for your all helping, that help me a lot.

shevshy

**Reply by burrbull** [reply]
There is a note in docs: https://docs.rs/stm32f4xx-hal/latest/stm32f4xx_hal/timer/delay/type.DelayMs.html

**Reply by burrbull** [reply, marked as answer]
> when I changed the delay_ms as delay_us. everything goes well. So, what the problem of that. Thanks a lot for your all helping, that help me a lot.

`delay_ms()` is just an alias for `delay::<1000>()` where `1000` is frequency of timer clocking used to calculate timer prescaler which is 16bit width.
84MHz / 1kHz == 85000 > 2^16
